### PR TITLE
PropagatedVersion changes to support namespaces

### PR DESCRIPTION
**Pull request by font**
- Changes to support`PropagatedVersion` for namespaces.
- Adds crudtester check for primary cluster so that we can run E2E tests with or without the host cluster being part of the federation.

---
This change is [Reviewable](https://reviewable.io/reviews/marun/federation-v2/49)

### CODE REVIEW

File: pkg/controller/sync/controller.go [hunk 1]
```diff
@@ -658,9 +669,11 @@ func updatePropagatedVersion(adapter federatedtypes.FederatedTypeAdapter,
 		if err != nil {
 			return err
 		}
+
```
Comment (marun) [inline pkg/controller/sync/controller.go]: (No action required) functional vs non-functional...

File: test/common/crudtester.go [hunk 2]
```diff
@@ -158,15 +163,14 @@ func (c *FederatedTypeCrudTester) CheckUpdate(template, placement, override pkgr
...
-	kind := adapter.Kind()
 
 	var initialAnnotation string
 	meta := adapter.ObjectMeta(template)
 	if meta.Annotations != nil {
 		initialAnnotation = meta.Annotations[AnnotationTestFederationCrudUpdate]
 	}
 
-	c.tl.Logf("Updating %s %q", kind, qualifiedName)
+	c.tl.Logf("Updating %s %q", c.kind, qualifiedName)
```
Comment (marun) [inline test/common/crudtester.go]: (No action required) Again, from a review perspective, this kind of change would ideally be separated out.

File: test/common/crudtester.go [hunk 3]
```diff
@@ -412,11 +422,20 @@ func (c *FederatedTypeCrudTester) waitForPropagatedVersion(template, placement,
 
 	clusterNames := c.adapter.Placement().ClusterNames(placement)
 	selectedClusters := sets.NewString(clusterNames...)
+	if c.kind == federatedtypes.NamespaceKind {
+		// Delete the primary cluster as it will never be included in
+		// PropagatedVersion's list of cluster verions.
```
Comment (marun) [inline test/common/crudtester.go]: s/verions/versions/

File: test/common/crudtester.go [hunk 4]
```diff
@@ -412,11 +422,20 @@ func (c *FederatedTypeCrudTester) waitForPropagatedVersion(template, placement,
...
+		selectedClusters.Delete(c.getPrimaryClusterName())
+	}
 
 	var version *fedv1a1.PropagatedVersion
 	err := wait.PollImmediate(c.waitInterval, c.clusterWaitTimeout, func() (bool, error) {
 		var err error
-		version, err = client.FederationV1alpha1().PropagatedVersions(namespace).Get(name, metav1.GetOptions{})
+		if c.kind == federatedtypes.NamespaceKind {
```
Comment (marun) [inline test/common/crudtester.go]: (No action required) Might make sense to add a helper for this pattern in a future PR.

File: test/common/crudtester.go [hunk 5]
```diff
@@ -451,7 +470,35 @@ func (c *FederatedTypeCrudTester) versionName(resourceName string) string {
...
+			continue
+		} else {
+			*clusterNames = append((*clusterNames)[:i], (*clusterNames)[i+1:]...)
+			break
+		}
+	}
+}
+
+func (c *FederatedTypeCrudTester) propagatedVersion(version *fedv1a1.PropagatedVersion, clusterName string) string {
+	// The primary cluster's namespace version is stored as the
```
Comment (marun) [inline test/common/crudtester.go]: Can you rephrase this?  It doesn't make sense to me as-is.

File: test/e2e/framework/unmanaged.go [hunk 6]
```diff
@@ -95,28 +97,33 @@ func (f *UnmanagedFramework) AfterEach() {
...
-	// Print events if the test failed and ran in a namep.
+	// Print events if the test failed and ran in a namespace.
 	if CurrentGinkgoTestDescription().Failed && f.testNamespaceName != "" {
 		kubeClient := f.KubeClient(userAgent)
 		DumpEventsInNamespace(func(opts metav1.ListOptions, ns string) (*corev1.EventList, error) {
 			return kubeClient.Core().Events(ns).List(opts)
 		}, f.testNamespaceName)
 	}
+
+	// DeleteNamespace at the very end but before tearing down the namespace
```
Comment (marun) [inline test/e2e/framework/unmanaged.go]: Cleanup needs to remain as a defer function because `DumpEventsInNamespace` contains assertions that could exit the function.

File: test/e2e/framework/unmanaged.go [hunk 7]
```diff
@@ -152,15 +159,29 @@ func (f *UnmanagedFramework) ClusterClients(userAgent string) map[string]kubecli
...
-	clusterClients := make(map[string]kubeclientset.Interface)
+	testClusters := make(map[string]common.TestCluster)
 	for _, cluster := range clusterList.Items {
 		ClusterIsReadyOrFail(fedClient, cluster)
 		config, err := util.BuildClusterConfig(&cluster, kubeClient, crClient)
 		Expect(err).NotTo(HaveOccurred())
 		restclient.AddUserAgent(config, userAgent)
-		clusterClients[cluster.Name] = kubeclientset.NewForConfigOrDie(config)
+
+		// Check if this cluster is the same name as the current-context to
```
Comment (marun) [inline test/e2e/framework/unmanaged.go]: Is this the best way to do this in the absence of a first-class cluster id?  I'm wondering if it might make sense to require a configuration parameter naming the host cluster instead.

File: test/e2e/framework/unmanaged.go [hunk 8]
```diff
@@ -152,15 +159,29 @@ func (f *UnmanagedFramework) ClusterClients(userAgent string) map[string]kubecli
...
+		// make it the primary cluster.
+		if cluster.Name == f.Kubeconfig.CurrentContext {
+			testClusters[cluster.Name] = common.TestCluster{
```
Comment (marun) [inline test/e2e/framework/unmanaged.go]: Also, consider simplifying: 

```golang
testClusters[cluster.Name] = common.TestCluster{
  kubeclientset.NewForConfigOrDie(config),
  (cluster.Name == hostClusterName),
}

File: test/common/crudtester.go (hunk 2 above)
Comment (font) [reply test/common/crudtester.go]: I definitely understand and mostly agree on your desire to have these things separated out from a review perspective. But it's too tedious to go back and undo each non-functional change from each individual commit, especially when a lot of this code was written during a mostly prototype phase. Moving forward, you can expect these things to be separated out.

File: test/common/crudtester.go (hunk 3 above)
Comment (font) [reply test/common/crudtester.go]: Fixed.

File: test/common/crudtester.go (hunk 5 above)
Comment (font) [reply test/common/crudtester.go]: Updated with the below phrasing. Let me know if that is more clear.
```
	// For namespaces, since we do not store the primary cluster's namespace
	// version in PropagatedVersion's ClusterVersions slice, grab it from the
	// TemplateVersion field instead.
```

File: test/e2e/framework/unmanaged.go (hunk 6 above)
Comment (font) [reply test/e2e/framework/unmanaged.go]: Ah, interesting. I added a comment about that and moved the original code, plus the fixture cleanup which depends on #48, to a defer function.
